OpenRA.Game/Graphics/TerrainSpriteLayer.cs

```diff
@@ -25,42 +25,70 @@ public sealed class TerrainSpriteLayer : IDisposable
 		readonly Sheet[] sheets;
 		readonly Sprite emptySprite;
 
+		readonly int binSize;
+		readonly int binCols;
+		readonly int binStride;
+
 		readonly IVertexBuffer<Vertex> vertexBuffer;
 		readonly Vertex[] vertices;
 		readonly bool[] ignoreTint;
-		readonly HashSet<int> dirtyRows = new HashSet<int>();
-		readonly int rowStride;
+		readonly bool[] dirtyBins;
+
 		readonly bool restrictToBounds;
 
 		readonly WorldRenderer worldRenderer;
 		readonly Map map;
 
 		readonly PaletteReference[] palettes;
 
-		public TerrainSpriteLayer(World world, WorldRenderer wr, Sprite emptySprite, BlendMode blendMode, bool restrictToBounds)
+		public TerrainSpriteLayer(World world, WorldRenderer wr, int binSize, Sprite emptySprite, BlendMode blendMode, bool restrictToBounds)
 		{
 			worldRenderer = wr;
+			this.binSize = binSize;
 			this.restrictToBounds = restrictToBounds;
 			this.emptySprite = emptySprite;
 			sheets = new Sheet[SpriteRenderer.SheetCount];
 			BlendMode = blendMode;
 
 			map = world.Map;
-			rowStride = 6 * map.MapSize.X;
+			binCols = Exts.IntegerDivisionRoundingAwayFromZero(map.MapSize.X, binSize);
+			var binRows = Exts.IntegerDivisionRoundingAwayFromZero(map.MapSize.Y, binSize);
+			binStride = 6 * binSize * binSize;
+			vertices = new Vertex[binRows * binCols * binStride];
 
-			vertices = new Vertex[rowStride * map.MapSize.Y];
 			palettes = new PaletteReference[map.MapSize.X * map.MapSize.Y];
 			vertexBuffer = Game.Renderer.Context.CreateVertexBuffer(vertices.Length);
 
+			dirtyBins = new bool[binRows * binCols];
+			Array.Fill(dirtyBins, true, 0, dirtyBins.Length);
+
 			wr.PaletteInvalidated += UpdatePaletteIndices;
 
 			if (wr.TerrainLighting != null)
 			{
-				ignoreTint = new bool[rowStride * map.MapSize.Y];
+				ignoreTint = new bool[vertices.Length];
 				wr.TerrainLighting.CellChanged += UpdateTint;
 			}
 		}
 
+		int VertexOffset(MPos uv)
+		{
+			var binCol = uv.U / binSize;
+			var binRow = uv.V / binSize;
+			var dx = uv.U - binCol * binSize;
+			var dy = uv.V - binRow * binSize;
+
+			return binStride * (binRow * binCols + binCol) + 6 * (dy * binSize + dx);
+		}
+
+		void MarkDirty(MPos uv)
+		{
+			var binCol = uv.U / binSize;
```

Comment:
Divides are expensive relative to addition, but additions and dives are both orders of magnitude faster than code that does any type of meaningful work. Do we have any benchmarks to show that avoiding divisions like this would have more than a 0.000001% performance impact?